pkg/operator/controllers/dnsmasq/machineconfigpool_controller_test.go

```diff
@@ -0,0 +1,254 @@
+package dnsmasq
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	mcv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	mcofake "github.com/openshift/machine-config-operator/pkg/generated/clientset/versioned/fake"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	arofake "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned/fake"
+	mock_dynamichelper "github.com/Azure/ARO-RP/pkg/util/mocks/dynamichelper"
+)
+
+func TestMachineConfigPoolReconciler(t *testing.T) {
+	fakeAro := func(objects ...kruntime.Object) *arofake.Clientset {
+		return arofake.NewSimpleClientset(objects...)
+	}
+	fakeMco := func(objects ...kruntime.Object) *mcofake.Clientset {
+		return mcofake.NewSimpleClientset(objects...)
+	}
+	fakeDh := func(controller *gomock.Controller) *mock_dynamichelper.MockInterface {
+		return mock_dynamichelper.NewMockInterface(controller)
+	}
+
+	tests := []struct {
+		name    string
+		arocli  *arofake.Clientset
+		mcocli  *mcofake.Clientset
+		mocks   func(mdh *mock_dynamichelper.MockInterface)
+		request ctrl.Request
+		wantErr bool
+	}{
+		{
+			name:    "no cluster",
+			arocli:  fakeAro(),
+			mcocli:  fakeMco(),
+			mocks:   func(mdh *mock_dynamichelper.MockInterface) {},
+			request: ctrl.Request{},
+			wantErr: true,
+		},
+		{
+			name: "controller disabled",
+			arocli: fakeAro(
+				&arov1alpha1.Cluster{
+					ObjectMeta: metav1.ObjectMeta{Name: "cluster"},
+					Status:     arov1alpha1.ClusterStatus{},
+					Spec: arov1alpha1.ClusterSpec{
+						OperatorFlags: arov1alpha1.OperatorFlags{
+							controllerEnabled: "false",
+						},
+					},
+				},
+			),
+			mcocli:  fakeMco(),
+			mocks:   func(mdh *mock_dynamichelper.MockInterface) {},
+			request: ctrl.Request{},
+			wantErr: false,
+		},
+		{
+			name: "no MachineConfigPool does nothing",
+			arocli: fakeAro(
+				&arov1alpha1.Cluster{
+					ObjectMeta: metav1.ObjectMeta{Name: "cluster"},
+					Status:     arov1alpha1.ClusterStatus{},
+					Spec: arov1alpha1.ClusterSpec{
+						OperatorFlags: arov1alpha1.OperatorFlags{
+							controllerEnabled: "true",
+						},
+					},
+				},
+			),
+			mcocli: fakeMco(),
+			mocks:  func(mdh *mock_dynamichelper.MockInterface) {},
+			request: ctrl.Request{
+				NamespacedName: types.NamespacedName{
+					Namespace: "",
+					Name:      "custom",
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "MachineConfigPool marked for deletion with no finalizers does nothing",
```

Comment:
why do we test this case? This can only happen if you delete the MCP before it was reconciled once, right?

Reply:
My understanding was that this scenario can also happen after the finalizer is handled and removed, but before the MCP itself gets deleted. If that is not the case, this test case and the corresponding guard check (machineconfigpool_controller.go#L69-71) could be removed.